xapian-core/queryparser/arabic-normalizer.cc

```diff
@@ -0,0 +1,122 @@
+    Xapian::Utf8Iterator i(word);
+    // FIXME order them by proba of happening
+    for (; i != Xapian::Utf8Iterator(); ++i) {
+        unsigned ch = *i, subst_ch = 0;
+        // convert letters to their unshaped form
+        if (normalize_shaping && 0 != (subst_ch = get_value(ch, ARABIC_SHAPING_MAP, LENGTH(ARABIC_SHAPING_MAP)))) {
+            ch = subst_ch;
+        }
+        // normalize LAM ALEF forms
+        if (normalize_shaping && FIND(ARABIC_LIGUATURES, ch)) {
+            Xapian::Unicode::append_utf8(new_word, ARABIC_LAM);
+            Xapian::Unicode::append_utf8(new_word, ARABIC_ALEF);
+        }
+        // strip diacritics
+        else if (normalize_diacritics && FIND(ARABIC_TASHKEEL, ch)) {
+            // ignore it
+        }
+        // strip the tatweel or kasheeda
+        else if (normalize_shaping && ch == ARABIC_TATWEEL) {
+            // ignore it
+        }
+        // strip arabic-specific punctuation
+        else if (normalize_punctuation && (ch == ARABIC_COMMA || ch == ARABIC_SEMICOLON || ch == ARABIC_QUESTION)) {
+            // ignore it
+        }
+        // convert TEH_MERBUTA to HEH
+        else if (normalize_similar_spellings && ch ==  ARABIC_TEH_MARBUTA) {
+            Xapian::Unicode::append_utf8(new_word, ARABIC_HEH);
+        }
+        // convert ALEF_MAKSURA to YEH
+        else if (normalize_similar_spellings && ch ==  ARABIC_ALEF_MAKSURA) {
+            Xapian::Unicode::append_utf8(new_word, ARABIC_YEH);
+        }
+        // normalize Alef forms
+        else if (normalize_similar_spellings && FIND(ARABIC_ALEFAT, ch)) {
+            Xapian::Unicode::append_utf8(new_word, ARABIC_ALEF);
+        }
+        // normalize Hamza forms
+        else if (normalize_hamza && FIND(ARABIC_HAMZAT, ch)) {
+            Xapian::Unicode::append_utf8(new_word, ARABIC_HAMZA);
+        }
+        // nothing to do
+        else {
+            Xapian::Unicode::append_utf8(new_word, ch);
+        }
+    }
+    return new_word;
+}
+
+string ArabicNormalizer::arabize(const string& word, int romanization_system) {
+    string new_word;
+    const character_map *mapping;
+    int mapping_length = 0;
+    Xapian::Utf8Iterator i(word);
+
+    if (romanization_system == 0) romanization_system = guess_romanization_system(word);
+    switch (romanization_system){
+        case 2: mapping = ISO233_TO_ARABIC; mapping_length = LENGTH(ISO233_TO_ARABIC); break;
+        /*case 1 is the default*/
+        default: mapping = BUCKWALTER_TO_ARABIC; mapping_length = LENGTH(BUCKWALTER_TO_ARABIC); break;
+    }
+
+    for (; i != Xapian::Utf8Iterator(); ++i) {
+        unsigned ch = *i, subst_ch = 0;
+        subst_ch = get_value(ch, mapping, mapping_length);
+        if ( subst_ch != 0 ) {
+            Xapian::Unicode::append_utf8(new_word, subst_ch);
+        }
+    }
+    return new_word;
+}
+
+int ArabicNormalizer::guess_romanization_system(const string& word) {
+    return 0 && LENGTH(word); // TODO to be implemented
```

Comment:
Each randomization has a special set of characters to use and specific rules. The idea is to use  this character to tighten the guessing interval till get to one system (see Comparison table https://en.wikipedia.org/wiki/Romanization_of_Arabic)